scalyr_agent/builtin_monitors/journald_monitor.py

```diff
@@ -0,0 +1,258 @@
+from scalyr_agent.monitor_utils.journald_checkpoints import Checkpoint
+import scalyr_agent.scalyr_logging as scalyr_logging
+
+try:
+    from systemd import journal
+except ImportError:
+    raise Exception('Python systemd library not installed.\n\nYou must install the systemd python library in order '
+                    'to use the journald monitor.\n\nThis can be done via package manager e.g.:\n\n'
+                    '  apt-get install python-systemd  (debian/ubuntu)\n'
+                    '  dnf install python-systemd  (CentOS/rhel/Fedora)\n\n'
+                    'or installed from source using pip e.g.\n\n'
+                    '  pip install systemd-python\n\n'
+                    'See here for more info: https://github.com/systemd/python-systemd/\n'
+                   )
+
+global_log = scalyr_logging.getLogger(__name__)
+__monitor__ = __name__
+
+define_config_option( __monitor__, 'journal_path',
+                     'Optional (defaults to /var/log/journal). Location on disk of the journald logs.',
+                     convert_to=str, default='/var/log/journal')
+
+define_config_option( __monitor__, 'journal_poll_timeout_milliseconds',
+                     'Optional (defaults to 0). The number of milliseconds to wait for data while polling the journal file. '
+                     'Note: you are better off setting the sample_interval monitor option if the timeout is longer than '
+                     'one second.',
+                     convert_to=int, default=0)
+
+define_config_option( __monitor__, 'journal_fields',
+                     'Optional dict containing a list of journal fields to include with each message, as well as a field name to map them to.\n'
+                     'Note: Not all fields need to exist in every message and only fields that exist will be included.\n'
+                     'Defaults to {:\n'
+                     '  "_SYSTEMD_UNIT": "unit"\n'
+                     '  "_PID": "pid"\n'
+                     '  "_MACHINE_ID": "machine_id"\n'
+                     '  "_BOOT_ID": "boot_id"\n'
+                     '  "_HOSTNAME": "hostname"\n'
+                     '  "_SOURCE_REALTIME_TIMESTAMP": timestamp"\n'
+                     '}\n',
+                     default=None
+                     )
+
+define_config_option( __monitor__, 'journal_matches',
+                      'Optional list containing match strings for filtering entries.  A match string consists of "FIELD=value" where FIELD '
+                      'is a field of the journal entry e.g. _SYSTEMD_UNIT, _HOSTNAME, _GID and "value" is the value to filter on. '
+                      'All matches of different field are combined with logical AND, and matches of the same field are automatically combined with logical OR. '
+                      'If empty or None then no filtering occurs.',
+                      default=None )
+
+define_config_option( __monitor__, 'checkpoint_name',
```

Comment:
I think we should do away with this and just rely on the existing monitor instance name mechanism to allow them to specify unique names for the monitor using the `id` configuration option.  If they don't specify an `id` then that's on them.

Reply:
There is no existing mechanism per se, instead it appears the `id` field is just a common convention used by some monitors (and each one implements as necessary).  I've updated the agent to also use id instead of checkpoint_name, or use the module_name if no id is specified.